modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,368 @@
+	if !cached {
+		mr.uncachedRoots[i] = root
+		return nil
+	}
+	// If the root was cached we need to rebuild the cache.
+	if err := mr.rebuildCachedTree(i); err != nil {
+		return errors.AddContext(err, "failed to rebuild cache for inserted root")
+	}
+	return nil
+}
+
+// isIndexCached determines if the root at index i is already cached in
+// mr.cachedSubTree or if it is still in mr.uncachedRoots. It will return true
+// or false and the index of the root in the corresponding data structure.
+func (mr *merkleRoots) isIndexCached(i int) (int, bool) {
+	if i/merkleRootsPerCache == len(mr.cachedSubTrees) {
+		// Root is not cached. Return the false and the position in
+		// mr.uncachedRoots
+		return i - len(mr.cachedSubTrees)*merkleRootsPerCache, false
+	}
+	return i / merkleRootsPerCache, true
+}
+
+// lenFromFile returns the number of merkle roots by computing it from the
+// filesize.
+func (mr *merkleRoots) lenFromFile() (int, error) {
+	stat, err := mr.file.Stat()
+	if err != nil {
+		return 0, err
+	}
+	size := stat.Size()
+	// If we haven't written a single root yet we just return 0.
+	if size < contractHeaderSize {
+		return 0, nil
+	}
+
+	// Sanity check contract file length.
+	if (size-contractHeaderSize)%crypto.HashSize != 0 {
+		return 0, errors.New("contract file has unexpected length and might be corrupted")
+	}
+	return int((size - contractHeaderSize) / crypto.HashSize), nil
+}
+
+// len returns the number of merkle roots. It should always return the same
+// number as lenFromFile.
+func (mr *merkleRoots) len() int {
+	return mr.numMerkleRoots
+}
+
+// appendRoot appends a root to the in-memory structure of the merkleRoots. If
+// the length of the uncachedRoots grows too large they will be compressed into
+// a cachedSubTree.
+func (mr *merkleRoots) appendRoot(root crypto.Hash) {
+	mr.uncachedRoots = append(mr.uncachedRoots, root)
+	if len(mr.uncachedRoots) == merkleRootsPerCache {
+		mr.cachedSubTrees = append(mr.cachedSubTrees, newCachedSubTree(mr.uncachedRoots))
+		mr.uncachedRoots = mr.uncachedRoots[:0]
+	}
+}
+
+// push appends a merkle root to the end of the contract. If the number of
+// uncached merkle roots grows too big we cache them in a new subTree.
+func (mr *merkleRoots) push(root crypto.Hash) error {
+	// Sanity check the number of uncached roots before adding a new one.
+	if len(mr.uncachedRoots) == merkleRootsPerCache {
+		build.Critical("the number of uncachedRoots is too big. They should've been cached by now")
+	}
+	// Calculate the root offset within the file and write it to disk.
+	rootOffset := fileOffsetFromRootIndex(mr.len())
+	if _, err := mr.file.WriteAt(root[:], rootOffset); err != nil {
+		return err
+	}
+	// Add the root to the unached roots.
+	mr.appendRoot(root)
+
+	// Increment the number of roots.
+	mr.numMerkleRoots++
+	return nil
+}
+
+// root returns the root of the merkle roots.
+func (mr *merkleRoots) root() crypto.Hash {
+	tree := crypto.NewCachedTree(sectorHeight)
```

Comment:
If we're working with pushing sub-roots, does it need to be a cached tree? Can it just be a normal tree?

I think that with the help of the `PushSubTree` function, we can stop using the `CachedTree` type.